### PR TITLE
 Support Licensee as a scanner

### PR DESCRIPTION
**Pull request by pnowakowski**
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/246)

### CODE REVIEW

**Comment by pnowakowski** [comment]
Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[scanner/src/funTest/kotlin/ScanPathTest.kt, line 58 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70WueJutj-x9T0e-r2-58:-L3cVjaAle3iFW0ZGJ9d:bxbhg05) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/funTest/kotlin/ScanPathTest.kt#L58)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Actually, this should be only `ExpensiveTag` (and I'll rename `ScanPathTag` to `ScanCodeTag` in a follow-up commit; sorry for the confusion).
</blockquote></details>

I'm not sure if Expensive is needed here, Licensee is fast.
Should we then remove this test to a separate class?

---

*[scanner/src/main/kotlin/Scanner.kt, line 40 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXF-r2-40:-L3cW9xssnv54CYAhrkO:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/Scanner.kt#L40)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Please sort this list alphabetically for a better overview (once it grows).
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 38 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-38:-L3cXE6S_3E0Dac_fyaA:b3f9d49) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L38)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Could we drop passing the working directory here and ...
</blockquote></details>

I'm affrid we can't..

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 40 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-40:-L3cXJpH9-hQ3Uo0yTiz:b-65g6mp) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L40)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

... in exchange use `path.absolutePath` here?
</blockquote></details>

... because it fails  with `-i path_to_single_file`  scanner input parameter  on Windows this way : 
```15:17:04.389 ERROR - Could not scan 'C:\Users\nowakow.pio\StudioProjects\oss-review-toolkit_licensee_support\LICENSE': 'licensee.bat C:\Users\nowakow.pio\StudioProjects\oss-review-toolkit_licensee_support\LICENSE' in directory 'C:\Users\nowakow.pio\StudioProjects\oss-review-toolkit_licensee_support' failed with exit code 1:
C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/licensee-9.8.0/lib/licensee/projects/fs_project.rb:22:in `initialize': Search root must be the project path directory or its ancestor (RuntimeError)
        from C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/licensee-9.8.0/lib/licensee.rb:44:in `new'
        from C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/licensee-9.8.0/lib/licensee.rb:44:in `rescue in project'
        from C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/licensee-9.8.0/lib/licensee.rb:37:in `project'
        from C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/licensee-9.8.0/bin/licensee:28:in `<top (required)>'
        from C:/Ruby25-x64/bin/licensee:23:in `load'
        from C:/Ruby25-x64/bin/licensee:23:in `<main>'
```

I tried to understand why it is like that: https://github.com/benbalter/licensee/blob/c39013111b4e0baf0c7c0100b8d7085c6c3b8bc6/lib/licensee/projects/fs_project.rb#L22
but my ability to read  ruby code was not good enough.  

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 45 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-45:-L3c_i5dtr_3odBst1JL:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L45)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Please use `stdoutFile.copyTo(resultsFile)` instead.
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 57 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-57:-L3c_lrvEgiuVqPp602J:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L57)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Nit: Blank line in front of this one.
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 58 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-58:-L3c_sjjnFE5a9PE8k22:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L58)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

How about this instead:

> Convert Licensee's output for "Closest licenses" (in case of non-exact matches) into proper YAML by replacing the asterisk with a dash.
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 59 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-59:-L3caAzT2NTeipk28Dq9:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L59)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

I believe we should use a dash (`-`) here instead of a blank string in order to make the similar licenses array entries in the parent "Closest licenses" field. Also, let's include the leading whitespace to match exactly that indentation level:

    val yamlResults = resultsFile.readText().replace("    * ", "    - ")
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 60 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-60:-L3cbo66ueCO0Qvq15_c:b-zezo1t) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L60)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Good idea in general to use the existing YAML reader to parse Licensee's output!
</blockquote></details>

Not sure if I understand what you mean. Do you mean somehow different use of `com.here.ort.utils.yamlMapper`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246)*

**Comment by sschuberth** [comment]
Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[scanner/src/funTest/kotlin/ScanPathTest.kt, line 58 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70WueJutj-x9T0e-r2-58:-L3ceF3xqBgnEfsI6Bd1:b-cwd33q) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/funTest/kotlin/ScanPathTest.kt#L58)):*
<details><summary><i>Previously, pnowakowski wrote…</i></summary><blockquote>

I'm not sure if Expensive is needed here, Licensee is fast.
Should we then remove this test to a separate class?
</blockquote></details>

No, then let's just use no tag at all for this particular test.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 60 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-60:-L3ceg921YdS2psP_XcE:bypm8ne) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L60)):*
<details><summary><i>Previously, pnowakowski wrote…</i></summary><blockquote>

Not sure if I understand what you mean. Do you mean somehow different use of `com.here.ort.utils.yamlMapper`?
</blockquote></details>

No, this was a compliment about your implementation :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246)*

**Comment by pnowakowski** [comment]
Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[scanner/src/funTest/kotlin/ScanPathTest.kt, line 58 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70WueJutj-x9T0e-r2-58:-L3cfgaupVeDaB0Z65k2:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/funTest/kotlin/ScanPathTest.kt#L58)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

No, then let's just use no tag at all for this particular test.
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 60 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-60:-L3cfknYxQ_VL7yy61WY:b-13vlvq) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L60)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

No, this was a compliment about your implementation :-)
</blockquote></details>

That was one of two the ways I understood it ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246)*

File: .travis.yml [hunk 1]
```diff
@@ -20,6 +20,7 @@ cache:
 
 install:
   - npm install -g npm@5.5.1 yarn@1.3.2
+  - gem install licensee
```
Comment (sschuberth) [inline .travis.yml]: Forgot to say, we should amend `-v 9.8.0` here to stick to a fixed version.

File: appveyor.yml [hunk 2]
```diff
@@ -21,6 +21,9 @@ install:
   - pip install virtualenv==15.1.0
   - cinst sbt --version 1.0.2 -y
   - refreshenv
+  - set PATH=C:\Ruby25\bin;%PATH%
+  - gem install licensee
```
Comment (sschuberth) [inline appveyor.yml]: Forgot to say, we should amend `-v 9.8.0` here to stick to a fixed version.

**Comment by sschuberth** [comment]
Reviewed 1 of 3 files at r2.
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 57 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-57:-L3cnhL4hYHMZf6BPDAu:bl8tpmv) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L57)):*
<details><summary><i>Previously, pnowakowski wrote…</i></summary><blockquote>

Done.
</blockquote></details>

*Before* the line with `if`, not after it :wink:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246)*

**Comment by pnowakowski** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[.travis.yml, line 23 at r4](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70WueJutj-x9T0c-r4-23:-L3g9SLPGTe3tFdy1TAq:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/2a25db72d52628d25151626334560d75a1640894/.travis.yml#L23)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Forgot to say, we should amend `-v 9.8.0` here to stick to a fixed version.
</blockquote></details>

Done.

---

*[appveyor.yml, line 25 at r4](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70WueJutj-x9T0d-r4-25:-L3g9TI2lH0KiKNyx9Pf:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/2a25db72d52628d25151626334560d75a1640894/appveyor.yml#L25)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Forgot to say, we should amend `-v 9.8.0` here to stick to a fixed version.
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/scanners/Licensee.kt, line 57 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-L3cG70VOjBCQhd3SOXG-r2-57:-L3gAFEgKy4pX6xZo8X_:bc63gk8) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/168622fe685fb0facca618d246b7ca7a48ad111d/scanner/src/main/kotlin/scanners/Licensee.kt#L57)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

*Before* the line with `if`, not after it :wink:
</blockquote></details>

So , wow it's done :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246)*

**Comment by sschuberth** [comment]
:lgtm:

---

Reviewed 1 of 5 files at r3, 4 of 4 files at r5.
Review status: 2 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246#-:-L3gI0gC55zTBLKZed4L:bnfp4nl)*

**Comment by sschuberth** [comment]
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/246)*

**Pull request merged**